### PR TITLE
Support type-use `@Nullable` annotation in annotated service 

### LINKED ISSUE #6454: `@Nullable` TYPE_USE annotation is not respected in annotated service
## Reproducible example

```java
public class FooService {

    @Get("/a")
    public String a(@Param @jakarta.annotation.Nullable String param1) {
        return "a"
    }

    @Get("/b")
    public String b(@Param @javax.annotation.Nullable String param2) {
        return "b";
    }

    @Get("/c")
    public String c(@Param @org.jspecify.annotations.Nullable String param3) {
        System.out.println("hi");
        return "c";
    }
}
```

```shell
$ curl localhost:8080/a
a%

$ curl localhost:8080/b
b%

$ curl localhost:8080/c
Status: 400
Description: Bad Request
```

## Description

This happens because checking `@Nullable` parameter does not consider `AnnotatedType`

https://github.com/line/armeria/blob/abc70dc2a24626f099a7ed80839c23e57f4ac78a/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java#L1025-L1033
- https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/AnnotatedType.html

I'm trying to apply jspecify `@Nullable` into armeria and this will be handled.

## Related issue
- https://github.com/line/centraldogma/pull/1194

### PR DESCRIPTION
**Pull request by sh-cho**
Motivation:

In annotated service, type-use `@Nullable` annotation (ex. jspecify) is not respected.

Modifications:

- Add jspecify dependency
  - (Migration in progress. See https://github.com/line/armeria/pull/6456)
- In `AnnotatedValueResolver#isAnnotatedNullable`, add checks for type-use nullable annotation

Result:

- Resolves #6454

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java [hunk 1]
```diff
@@ -1023,12 +1024,41 @@ private static Type parameterizedTypeOf(AnnotatedElement element) {
     }
 
     static boolean isAnnotatedNullable(AnnotatedElement annotatedElement) {
+        // 1) declaration annotation
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:1044]: Could you add examples so that readers can distinguish them easily? e.g.
```
@Nullable // declaration annotation
public String declarationAnnotatedMethod() {
    return null;
}

// type-use annotation
public @Nullable String typeUseAnnotatedMethod() {
    return null;
}
```

File: core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceNullableParamTest.java [hunk 2]
```diff
@@ -98,20 +103,29 @@ public String defaultValue(@Header @Default("unspecified") String value) {
                 public String optional(@Header Optional<String> value) {
                     return value.orElse("unspecified");
                 }
+
+                @Get("/type_use_nullable")
+                public String typeUseNullable(@Header @org.jspecify.annotations.Nullable String value) {
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceNullableParamTest.java:108]: Could you also add a test with this?
```
public @Nullable String typeUseAnnotatedMethod() {
    return null;
}
```

**Review by minwoox — APPROVED** [review]
Left small suggestions. Thanks, @sh-cho!

File: core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceNullableParamTest.java (hunk 2 above)
Comment (ikhoon) [reply core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceNullableParamTest.java:108]: We may use `@RequestObject` to test it.

File: core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java (hunk 1 above)
Comment (sh-cho) [reply core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java:1044]: Ok, I added comment
[`ca1e429` (#6457)](https://github.com/line/armeria/pull/6457/commits/ca1e429a8a566d9ba8c174f130b55c17807ae254)

File: core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceNullableParamTest.java (hunk 2 above)
Comment (sh-cho) [reply core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceNullableParamTest.java:108]: Added test for nullable response
[`8d850d9` (#6457)](https://github.com/line/armeria/pull/6457/commits/8d850d9e7b5a00190af30634a0fffd4a3c02625f)

Also I tried to test with `@RequestObject` like below, but looks like `@Nullable` doesn't affect 👀   (empty body is rejected even if there's `@Nullable`)

```java
@Post("/foo")
@ProducesJson
public Map<String, Object> foo(@RequestObject Map<String, Object> map) {
    return map;
}

@Post("/bar")
@ProducesJson
public Map<String, Object> bar(@RequestObject @Nullable Map<String, Object> map) {
    if (map == null) {
        return Map.of("this is", "null");
    }
    return map;
}
```

Comment (minwoox) [reply core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceNullableParamTest.java:108]: > empty body is rejected even if there's @Nullable

I think it shouldn't throw an exception if it's annotated with the `@Nullable`.
What do you think of it? @line/dx

Comment (jrhee17) [reply core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceNullableParamTest.java:108]: I agree the behavior sounds reasonable - I'm not sure whether this is enforceable for all cases though. I have no objection if for Jackson's case where a literal `null` body is received, a null value is passed as the request object.

Having said this, I'm fine with going ahead with merging this PR and handling this separately as they seem to be separate issues

Comment (minwoox) [reply core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceNullableParamTest.java:108]: > I'm fine with going ahead with merging this PR and handling this separately as they seem to be separate issues

I agree. We don't have to deal with this in this PR. Let me merge this. 😉

**Comment by minwoox** [comment]
Thanks a lot, @sh-cho! 😄
